src/devices/Mcp3xxx/Mcp3xxx.cs

```diff
@@ -0,0 +1,158 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Device.Spi;
+
+namespace Iot.Device.Adc
+{
+    /// <summary>
+    /// Mcp3xxx Abstract class representing the MCP 10 and 12 bit ADC devices.
+    /// </summary>
+    public abstract class Mcp3xxx : Mcp3Base
+    {
+        /// <summary>
+        /// the number of single ended input channel on the ADC
+        /// </summary>
+        protected byte ChannelCount;
+
+        private byte _adcResolutionBits;
+
+        /// <summary>
+        /// Constructs Mcp3xxx instance
+        /// </summary>
+        /// <param name="spiDevice">Device used for SPI communication</param>
+        /// <param name="channelCount">Value representing the number of single ended input channels available on the device.</param>
+        /// <param name="adcResolutionBits">The number of bits of resolution for the ADC.</param>
+        public Mcp3xxx(SpiDevice spiDevice, byte channelCount, byte adcResolutionBits) : base(spiDevice)
+        {
+            ChannelCount = channelCount;
+            _adcResolutionBits = adcResolutionBits;
+        }
+
+        /// <summary>
+        /// Checks that the channel is in range of the available channels channels and throws an exception if not.
+        /// </summary>
+        /// <param name="channel">Channel to be checked</param>
+        /// <param name="channelCount">Value representing the number of channels on the device which may vary depending on the configuration.</param>
+        protected void CheckChannelRange(int channel, int channelCount)
+        {
+            if (channel < 0 || channel > channelCount - 1)
+            {
+                throw new ArgumentOutOfRangeException($"ADC channel must be within the range 0-{channelCount - 1}.", nameof(channel));
+            }
+        }
+
+        /// <summary>
+        /// Checks that the channel is in range of the available input channels and that both channels are part of a valid pairing of input channels.
+        /// </summary>
+        /// <param name="valueChannel">Value channel to be checked</param>
+        /// <param name="referenceChannel">Reference channel to be checked</param>
+        protected void CheckChannelPairing(int valueChannel, int referenceChannel)
+        {
+            CheckChannelRange(valueChannel, ChannelCount);
+            CheckChannelRange(referenceChannel, ChannelCount);
+
+            // Check that the channels are in the differential pairing.
+            // When using differential inputs then then the single ended inputs are grouped into channel pairs such that for an 8 input deviice then the pairs
+            // would be CH0 and CH1, CH2 and CH3, CH4 and CH5, CH6 and CH7 and thus to work out which channel pairing a channel is in then the channel number can be divided by 2.
+            if (valueChannel / 2 != referenceChannel / 2 || valueChannel == referenceChannel)
```

Comment:
would that be equivalent to `Math.Abs(valueChannel  - referenceChannel) != 1`?

Reply:
Nope because if we use The channels 3 and 4 then it will pass the test but that are in different pairings. Another way would be to compare the values with bit 0 masked out.